### PR TITLE
Fix clone() bug.

### PR DESCRIPTION
**Pull request by gongweibao**
Fix https://github.com/PaddlePaddle/Paddle/issues/12628

### CODE REVIEW

File: python/paddle/fluid/framework.py [hunk 1]
```diff
@@ -1300,6 +1439,26 @@ def __init__(self):
         self._current_role = core.op_proto_and_checker_maker.OpRole.Forward
         self._op_role_var = []
 
+    def equal(self, other):
```
Comment (reyoung) [inline python/paddle/fluid/framework.py]: Please use Python style `__eq__` method

https://docs.python.org/3/reference/datamodel.html#object.__eq__

or move this method to unittest?

**Review by reyoung — CHANGES_REQUESTED** [review]
It seems that `self.attrs` is unnecessary.

File: paddle/fluid/framework/op_desc.cc [hunk 2]
```diff
@@ -238,6 +238,19 @@ Attribute OpDesc::GetNullableAttr(const std::string &name) const {
   }
 }
 
+std::vector<int> OpDesc::GetBlocksAttr(const std::string &name) const {
```
Comment (panyx0718) [inline paddle/fluid/framework/op_desc.cc]: GetBlocksID? not GetBlocksAttr.

File: python/paddle/fluid/framework.py [hunk 3]
```diff
@@ -574,6 +574,74 @@ def find_name(var_list, name):
...
+                    raise ValueError(
+                        "In Operator(Object) not equal:{0}\n".format(k))
+                    return False
+                continue
+
+            if k == 'attrs':
+                if len(self.attrs) != len(other.attrs):
+                    raise ValueError(
+                        "In Operator(Object)'s attrs's number not equal\n")
+                    return False
```
Comment (panyx0718) [inline python/paddle/fluid/framework.py]: no need?

**Comment by panyx0718** [comment]
Can you add more description of the PR?

What bug does it cause and why?

File: python/paddle/fluid/framework.py [hunk 4]
```diff
@@ -574,6 +574,74 @@ def find_name(var_list, name):
             self.desc.infer_var_type(self.block.desc)
             self.desc.infer_shape(self.block.desc)
 
+    def equal(self, other):
+        for k, v in self.__dict__.iteritems():
+            if isinstance(v, core.OpDesc) or isinstance(
+                    v, Program) or isinstance(v, Block):
+                continue
```
Comment (panyx0718) [inline python/paddle/fluid/framework.py]: can you favor if-else over if-continue? It's hard to know the `if`s you added is logically complete.

File: python/paddle/fluid/framework.py [hunk 5]
```diff
@@ -574,6 +574,74 @@ def find_name(var_list, name):
             self.desc.infer_var_type(self.block.desc)
             self.desc.infer_shape(self.block.desc)
 
+    def equal(self, other):
+        for k, v in self.__dict__.iteritems():
+            if isinstance(v, core.OpDesc) or isinstance(
```
Comment (panyx0718) [inline python/paddle/fluid/framework.py]: why is OpDesc not compared?

File: paddle/fluid/framework/op_desc.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/framework/op_desc.cc]: Yep. I'm fixing it!

File: python/paddle/fluid/framework.py (hunk 1 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: OK!Thanks!

File: python/paddle/fluid/framework.py (hunk 5 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: Done.

File: python/paddle/fluid/framework.py (hunk 4 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: Done.

File: python/paddle/fluid/framework.py (hunk 3 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: Done.

**Review by reyoung — DISMISSED** [review]

File: paddle/fluid/API.spec [hunk 6]
```diff
@@ -18,6 +18,9 @@ paddle.fluid.Operator.all_attrs ArgSpec(args=['self'], varargs=None, keywords=No
 paddle.fluid.Operator.attr ArgSpec(args=['self', 'name'], varargs=None, keywords=None, defaults=None)
 paddle.fluid.Operator.attr_type ArgSpec(args=['self', 'name'], varargs=None, keywords=None, defaults=None)
 paddle.fluid.Operator.block_attr ArgSpec(args=['self', 'name'], varargs=None, keywords=None, defaults=None)
+paddle.fluid.Operator.block_attr_id ArgSpec(args=['self', 'name'], varargs=None, keywords=None, defaults=None)
```
Comment (typhoonzero) [inline paddle/fluid/API.spec:21]: `block_attr` API is going to be hide by https://github.com/PaddlePaddle/Paddle/pull/12543, and these APIs should not be exposed. Please update this PR when #12543 is merged.

File: python/paddle/fluid/framework.py [hunk 7]
```diff
@@ -774,32 +774,84 @@ def attr(self, name):
...
+
+        return self.desc.blocks_attr_ids(name)
 
     def all_attrs(self):
         """
         Get the attribute dict.
 
         Returns:
-            dict: The Operator's attribute dict.
+            attr_map(dict(name->attr)): The Operator's attribute dict.
```
Comment (typhoonzero) [inline python/paddle/fluid/framework.py]: See: http://www.sphinx-doc.org/en/stable/ext/example_google.html docstring return comments better to be like the format:

```
    Returns:
    type: comment
```

File: python/paddle/fluid/initializer.py [hunk 8]
```diff
@@ -264,7 +264,8 @@ def __call__(self, var, block):
                 "dtype": int(var.dtype),
                 "mean": self._mean,
                 "std": self._std_dev,
-                "seed": self._seed
+                "seed": self._seed,
+                "use_mkldnn": False
```
Comment (typhoonzero) [inline python/paddle/fluid/initializer.py:268]: This line is intended for some other fixes?

Comment (gongweibao) [reply python/paddle/fluid/initializer.py:268]: 为了比较相等。
attr c++端有默认属性，python端不设置就不好比较，除非硬编码。

File: paddle/fluid/API.spec (hunk 6 above)
Comment (gongweibao) [reply paddle/fluid/API.spec:21]: OK.I'll fix them.

File: python/paddle/fluid/framework.py (hunk 7 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: Don.Thanks!

File: python/paddle/fluid/framework.py [hunk 9]
```diff
@@ -774,32 +774,84 @@ def attr(self, name):
...
 
+    def block_attr_id(self, name):
+        """
+        Get the block attribute's id by name.
+
+        Args:
+            name(str): the attribute name.
+
+        Returns:
+            (int): the block index.
```
Comment (typhoonzero) [inline python/paddle/fluid/framework.py]: This line also need to be updated.

File: python/paddle/fluid/framework.py [hunk 10]
```diff
@@ -774,32 +774,84 @@ def attr(self, name):
...
     def block_attr(self, name):
         """
-        Get the block attribute by name.
+        Get the block attribute  by name.
+
+        Args:
+            name(str): the attribute name.
+
+        Returns:
+            (block): the block attribute.
```
Comment (typhoonzero) [inline python/paddle/fluid/framework.py]: This line also need to be updated.

File: python/paddle/fluid/framework.py [hunk 11]
```diff
@@ -774,32 +774,84 @@ def attr(self, name):
...
+
+    def blocks_attr_ids(self, name):
+        """
+        Get the blocks attribute's ids by name.
+
+        Args:
+            name(str): the attribute name.
+
+        Returns:
+            (list[int]): the blocks index.
```
Comment (typhoonzero) [inline python/paddle/fluid/framework.py]: This line also need to be updated.

File: python/paddle/fluid/framework.py (hunk 9 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: Done.

File: python/paddle/fluid/framework.py (hunk 10 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: Done.

File: python/paddle/fluid/framework.py (hunk 11 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py]: Done.

File: python/paddle/fluid/tests/unittests/test_dist_base.py [hunk 12]
```diff
@@ -36,10 +36,18 @@ def start_pserver(self, model_file):
             (self._python_interp, model_file, self._ps_endpoints, ps1_ep,
              self._trainers)
 
+        ps0_pipe = subprocess.PIPE
+        ps1_pipe = subprocess.PIPE
+        if check_error_log:
+            print("ps0_cmd:", ps0_cmd)
+            print("ps1_cmd:", ps1_cmd)
+            ps0_pipe = open("/tmp/ps0_err.log", "wb")
```
Comment (typhoonzero) [inline python/paddle/fluid/tests/unittests/test_dist_base.py:147]: `ps0_pipe` these handles never closed

Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/test_dist_base.py:147]: Done. Thanks.

Comment (typhoonzero) [reply python/paddle/fluid/tests/unittests/test_dist_base.py:147]: please resolve the conflicts thanks.

Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/test_dist_base.py:147]: Done.

**Review by typhoonzero — APPROVED** [review]

**Review by reyoung — COMMENTED** [review]
Excellent

**Review by reyoung — APPROVED** [review]

**Pull request merged**